core/domain/exp_jobs_one_off.py

```diff
@@ -116,35 +118,85 @@ def reduce(key, values):
         yield (key, values)
 
 
-class MathExpressionInputInteractionOneOffJob(jobs.BaseMapReduceOneOffJobManager):  # pylint: disable=line-too-long
-    """Job that produces a list of (exploration, state) pairs that use the Math
-    Expression Interaction and that have rules that contain [<, >, =] to
-    identify and prevent usage of equation/inequalities.
+class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of explorations that use the MathExpressionInput
+    along with the validity and type (expression/equation) of the inputs present
+    in the exploration.
+
+    This validation is done by the 'is_valid_math_expression' or the
+    'is_valid_math_equation' function present in schema_utils.py.
     """
 
+    # This threshold puts a cap on the number of valid inputs, i.e.,
+    # expressions/equations that can be yielded by this one-off job.
+    VALID_MATH_INPUTS_YIELD_LIMIT = 200
+    UNICODE_TO_TEXT = {
+        u'\u221a': 'sqrt',
+        u'\u03c0': 'pi',
+        u'\xb7': '*'
+    }
+
     @classmethod
     def entity_classes_to_map_over(cls):
         return [exp_models.ExplorationModel]
 
     @staticmethod
     def map(item):
+        is_valid_math_expression = schema_utils.get_validator(
+            'is_valid_math_expression')
+        is_valid_math_equation = schema_utils.get_validator(
+            'is_valid_math_equation')
+        ltt = latex2text.LatexNodes2Text()
+        unicode_to_text_mapping = (
+            MathExpressionValidationOneOffJob.UNICODE_TO_TEXT)
+
         if not item.deleted:
             exploration = exp_fetchers.get_exploration_from_model(item)
             for state_name, state in exploration.states.items():
                 if state.interaction.id == 'MathExpressionInput':
                     for group in state.interaction.answer_groups:
                         for rule_spec in group.rule_specs:
-                            rule = rule_spec.inputs['x']
-                            if any(sep in rule for sep in SEPARATORS):
-                                yield (
-                                    item.id,
-                                    '%s: %s' % (
-                                        state_name.encode('utf-8'),
-                                        rule.encode('utf-8')))
+                            rule_input = ltt.latex_to_text(
+                                rule_spec.inputs['x'])
+
+                            # The pylatexenc lib outputs the unicode values of
+                            # special characters like sqrt and pi, which is why
+                            # they need to be replaced with their corresponding
+                            # text values before performing validation.
+                            for unicode_char, text in (
```

Comment:
The typical way to handle this sort of thing is to use .encode('utf-8'). You'll see this in existing tests. Why don't you do something like that instead, rather than all this special-casing?

Reply:
The purpose of doing this is to replace special characters' unicode values like u221a for sqrt with their string representations. So for instance, `\sqrt{2}` will get converted to `\u221a(2)` by pylatexenc. So, after replacing with the corresponding text value, we'll get `sqrt(2)` which is exactly what we want.

Reply:
Oh, I see, you need it for is_valid_math_expression().

OK then can you apply the .encode() thing instead of lines 171-179 below? That is how we typically handle "out of range" chars.

Reply:
@seanlip It's not only for the validation functions. Even while yielding the result, it is required to not have characters with ordinal values greater than 127. Using `.encode()` gives the following error:

```
File "/home/prayush/Desktop/opensource/oppia/core/domain/exp_jobs_one_off.py", line 192, in map
    item.id, state_name.encode('utf-8'), rule_input.encode('utf-8')))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 0: ordinal not in range(128)
```

Reply:
Can you check where the encode() error is happening -- is it in the state name or the rule input?

What is an example of an input causes this -- can you walk me through a trace of how it goes through the various stages of your map()?

Reply:
It's happening due to the rule input. I'll explain this through the example I've added to the tests. Let's assume the rule input is `âéîôü` (contains characters that have ordinal values > 127). If these characters are from the ones that we are expecting, like the square root sign or the multiplication dot, then we simply replace them with the corresponding strings (`√` -> `sqrt`) and everything is fine. But, if the input contains characters that we are not expecting, they wouldn't get replaced and these "out of range" characters would still be present in the rule_input string. So, when we try to perform some operation on this string (during validation) or when they are yielded by the map function, the string gets decoded but due to these "out of range" characters, a UnicodeDecodeError is thrown. To avoid this, what I've done is, I've replaced all characters that have ordinal values > 127 with their corresponding ordinal values. This is essentially the same as replacing square root's character with `sqrt`, but since we can't special-case _all_ possible characters, we just replace them with the ordinal value. So, the string `âéîôü` gets converted to `226 233 238 244 252` which does not contain characters that are out of range and everything works smoothly. Also, note that that particular input is going to be marked as invalid anyways(since it contains unexpected characters) so we can modify it without affecting the validation process.

Reply:
Tried it out. Still no good, I'm afraid -- the same errors still arise.

I would encourage you to take a closer look at encode('utf-8'). We've run into unicode issues before and they are often due to the MapReduce pipeline handling strings rather than unicode. It seems surprising to me that you don't need them for this job and you need to roll your own custom implementation. You might want to look at the patterns in other jobs and compare and contrast them to what you are doing here.

In general, try to avoid doing custom handcrafted things. Figure out what the output of pylatexenc is (str/unicode/etc.) and encode/decode it as needed using a standard method so that you can have guarantees on the types/encoding of your strings. Feel free to add assert statements/checks within the function to verify those guarantees -- if any of those assertions errors out on a production run, I will let you know.

Also, test thoroughly locally with unicode strings in various places. The error may not occur where you think it does. I suspect it's occurring because you're passing unicode chars to the MR framework somehow when it's expecting a string.

Reply:
Sorry about this, @seanlip. I have removed the customized replacements for out of range characters and used `.encode(utf-8')` instead. I think the errors are occurring because the yield string is in unicode and as you mentioned, they need to be strings. I've tested the latest changes locally and they do work -- please let me know if there's a specific way that I can test them to ensure that they would work on production.